org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend

```diff
@@ -119,31 +119,36 @@ import org.eclipse.lsp4j.Range
     }
 
     def String getSubstring(Range range) {
-    		val start = getOffSet(range.start)
-    		val end = getOffSet(range.end)
-    		return this.contents.substring(start, end)
+          val start = getOffSet(range.start)
+          val end = getOffSet(range.end)
+          return this.contents.substring(start, end)
     }
 
     def Document applyChanges(Iterable<? extends TextEdit> changes) {
-        var newContent = contents
+        var currentDocument = this
         for (change : changes) {
-            if (change.range === null) {
-                newContent = change.newText
-            } else {
-                val start = getOffSet(change.range.start)
-                val end = getOffSet(change.range.end)
-                newContent = newContent.substring(0, start) + change.newText + newContent.substring(end)
-            }
+            val newContent = if (change.range === null) {
```

Comment:
I wonder if this is correct, given the spec of the lsp:
```
TextEdit[]
Complex text manipulations are described with an array of TextEdit’s, representing a single change to the document.

All text edits ranges refer to positions in the original document. Text edits ranges must never overlap, that means no part of the original document must be manipulated by more than one edit. However, it is possible that multiple edits have the same start position: multiple inserts, or any number of inserts followed by a single remove or replace edit. If multiple inserts have the same position, the order in the array defines the order in which the inserted strings appear in the resulting text.
```

Reply:
TextDocumentContentChangeEvent do seem different from TextEdits.
i find it bogus that the semantics is different here

Reply:
=> @JanKoehnlein do you know why we use the same code for text edits and contentChanges